stl/inc/memory

```diff
@@ -2416,26 +2437,283 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_strong_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
+#if _HAS_CXX20 && !defined(_M_CEE_PURE)
+template <class _Ty>
+struct atomic<shared_ptr<_Ty>> : _Atomic_storage<shared_ptr<_Ty>> {
+private:
+    using _Base = _Atomic_storage<shared_ptr<_Ty>>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+#if 1 // TRANSITION, ABI
+    static constexpr bool is_always_lock_free =
+        sizeof(shared_ptr<_Ty>) <= 8 && (sizeof(shared_ptr<_Ty>) & sizeof(shared_ptr<_Ty>) - 1) == 0;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return is_always_lock_free;
+    }
+#else // ^^^ don't break ABI / break ABI vvv
+#if _ATOMIC_HAS_DCAS
+    static constexpr bool is_always_lock_free = sizeof(shared_ptr<_Ty>) <= 2 * sizeof(void*);
+#else // ^^^ _ATOMIC_HAS_DCAS / !_ATOMIC_HAS_DCAS vvv
+    static constexpr bool is_always_lock_free = sizeof(shared_ptr<_Ty>) <= sizeof(void*);
+#endif // _ATOMIC_HAS_DCAS
+
+    _NODISCARD bool is_lock_free() const noexcept {
+#if _ATOMIC_HAS_DCAS
+        return is_always_lock_free;
+#else // ^^^ _ATOMIC_HAS_DCAS / !_ATOMIC_HAS_DCAS vvv
+        return is_always_lock_free || (sizeof(shared_ptr<_Ty>) <= 2 * sizeof(void*) && __std_atomic_has_cmpxchg16b());
+#endif // _ATOMIC_HAS_DCAS
+    }
+#endif // TRANSITION, ABI
+
+    void store(shared_ptr<_Ty> _Value) noexcept {
+        shared_ptr<_Ty> _Old = exchange(_STD move(_Value));
+    }
+
+    void store(shared_ptr<_Ty> _Value, const memory_order _Order) noexcept {
+        shared_ptr<_Ty> _Old = exchange(_STD move(_Value), _Order);
+    }
+
+    _NODISCARD shared_ptr<_Ty> load() const noexcept {
+        shared_ptr<_Ty> _Result = _Base::load();
+        _Result._Incref();
+        return _Result;
+    }
+
+    _NODISCARD shared_ptr<_Ty> load(const memory_order _Order) const noexcept {
+        shared_ptr<_Ty> _Result = _Base::load(_Order);
+        _Result._Incref();
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(shared_ptr<_Ty> _Value) noexcept {
+        _Value._Incref();
+        return _Base::exchange(_STD move(_Value));
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(shared_ptr<_Ty> _Value, const memory_order _Order) noexcept {
+        _Value._Incref();
+        return _Base::exchange(_STD move(_Value), _Order);
+    }

```

Comment:
why can't this be the only implementation of exchange?

Reply:
the main atomic template adds 2 versions of each function so I've done the same thing. Just trying to be consistent. But I can change that if you really want.